rhp/v4/server.go

```diff
@@ -0,0 +1,1078 @@
+
+	// add the renter's parents to our transaction pool to ensure they are valid
+	// and update the proofs.
+	if len(req.RenterParents) > 0 {
+		if _, err := s.chain.AddV2PoolTransactions(req.Basis, req.RenterParents); err != nil {
+			return errorBadRequest("failed to add formation parents to transaction pool: %v", err)
+		}
+	}
+
+	// add the setup parents to the transaction pool
+	if len(renewalParents) > 0 {
+		if _, err := s.chain.AddV2PoolTransactions(basis, renewalParents); err != nil {
+			return errorBadRequest("failed to add setup parents to transaction pool: %v", err)
+		}
+	}
+
+	// get the full updated transaction set for the renewal transaction
+	basis, renewalSet, err := s.chain.V2TransactionSet(basis, renewalTxn)
+	if err != nil {
+		return fmt.Errorf("failed to get transaction set: %w", err)
+	} else if _, err = s.chain.AddV2PoolTransactions(basis, renewalSet); err != nil {
+		return errorBadRequest("failed to broadcast setup transaction: %v", err)
+	}
+	// broadcast the transaction set
+	s.syncer.BroadcastV2TransactionSet(basis, renewalSet)
+
+	// add the contract to the contractor
+	err = s.contractor.RenewV2Contract(TransactionSet{
+		Transactions: renewalSet,
+		Basis:        basis,
+	}, Usage{
+		RPCRevenue:       prices.ContractPrice,
+		StorageRevenue:   renewal.NewContract.HostOutput.Value.Sub(renewal.NewContract.TotalCollateral).Sub(prices.ContractPrice),
+		RiskedCollateral: renewal.NewContract.TotalCollateral.Sub(renewal.NewContract.MissedHostValue),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to add contract: %w", err)
+	}
+
+	// send the finalized transaction set to the renter
+	return rhp4.WriteResponse(stream, &rhp4.RPCRenewContractThirdResponse{
+		Basis:          basis,
+		TransactionSet: renewalSet,
+	})
+}
+
+func (s *Server) handleRPCVerifySector(stream net.Conn) error {
+	var req rhp4.RPCVerifySectorRequest
+	if err := rhp4.ReadRequest(stream, &req); err != nil {
+		return errorDecodingError("failed to read request: %v", err)
+	} else if err := req.Validate(); err != nil {
+		return rhp4.NewRPCError(rhp4.ErrorCodeBadRequest, err.Error())
+	}
+
+	prices := req.Prices
+	if err := prices.Validate(s.hostKey.PublicKey()); err != nil {
+		return fmt.Errorf("failed to validate price table: %w", err)
+	}
+	token := req.Token
+	if err := token.Validate(); err != nil {
+		return fmt.Errorf("failed to validate token: %w", err)
+	}
+
+	if err := s.contractor.DebitAccount(token.Account, prices.RPCVerifySectorCost()); err != nil {
+		return fmt.Errorf("failed to debit account: %w", err)
+	}
+
+	sector, err := s.sectors.ReadSector(req.Root)
+	if err != nil {
+		return rhp4.NewRPCError(rhp4.ErrorCodeBadRequest, err.Error())
+	}
+
+	// TODO: build proof
+	resp := rhp4.RPCVerifySectorResponse{
+		Leaf: ([64]byte)(sector[rhp4.LeafSize*req.LeafIndex:]),
+	}
+	return rhp4.WriteResponse(stream, &resp)
+}
+
+func updateStateElementBasis(cm ChainManager, base, target types.ChainIndex, element *types.StateElement) error {
```

Comment:
`ChainManager` has a method for this, right? But IIRC it expects a full `types.V2Transaction`. Seems to me like we should expose that somehow instead of defining `updateStateElementBasis` and `updateSiacoinElementBasis` here.

Reply:
IIRC only `AddV2PoolTransaction` updates the proofs. The problem that you run into is the host and renter are passing elements around with proofs at various heights. I'm fine with moving this into the ChainManager. It feels like a missed opportunity that each state element can't communicate how up-to-date its proof is.

Reply:
Like `types.StateElement` would have a `basis ChainIndex` field? That feels excessive to me, given that in most contexts, every basis in a transaction/block will be identical. It's only relevant when communicating with peers -- which is why we make it explicit in RHPv4 (as well as `AddV2PoolTransaction`).

Seems like just exporting `(ChainManager).updateV2Transaction` would do the trick. Taking a step back, though: do we actually need to be updating the elements here? If the renter's elements aren't up-to-date, `AddV2PoolTransaction` will update them; if they're *too* outdated, it will reject them, and we'll refuse to send our signatures to the renter.

Reply:
Will it not cause problems if we call Update on a transaction when the host's elements are `basis A` and the renter's elements are `basis B`?

Reply:
ah, I see. Yeah, you can't mix-and-match bases, all the elements have to have the same basis. So we need a method like:
```go
func (ChainManager) ChangeV2TransactionSetBasis(txns []types.V2Transaction, from, to types.ChainIndex) error
```
Then the process is:
1. Create a txn set with some basis
2. Receive elements from peer with some other basis
3. Change peer elements to your basis
4. Submit to txpool (which may do another update, if a block has been mined since you created the txn set)

Reply:
We don't get the actual transaction from the peer in this case, but maybe we should just build a temp transaction so that it can be a bit more generic?